File: backend/app/schemas/role.py
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict

from .common import Permission


class Role(BaseModel):
    """Basic role information"""
    id: int
    name: str
    description: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class RoleCreate(BaseModel):
    """Schema for creating a new role"""
    name: str = Field(..., min_length=1, max_length=50)
    description: Optional[str] = Field(None, min_length=1, max_length=200)


class RoleUpdate(BaseModel):
    """Schema for updating a role"""
    name: Optional[str] = Field(None, min_length=1, max_length=50)
    description: Optional[str] = Field(None, min_length=1, max_length=200)


class RoleDetail(BaseModel):
    """Schema for detailed role information"""
    id: int
    name: str
    description: Optional[str] = None
    permissions: Optional[List[Permission]] = []
    user_count: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/types/role.ts
import type { UUID, Permission } from './common';

/**
 * Role type definitions
 * These types match the backend Pydantic schemas for Role-related operations
 */

export interface Role {
  id: UUID;
  name: string;
  description: string;
  hierarchy_order: number;
  created_at: string;
  updated_at: string;
}

export interface RoleDetail {
  id: UUID;
  name: string;
  description: string;
  created_at: string;
  updated_at: string;
  permissions: Permission[];
  user_count?: number;
}

export interface RoleCreate {
  name: string;
  description: string;
  hierarchy_order?: number;
}

export interface RoleUpdate {
  name?: string;
  description?: string;
}

export interface RoleReorderItem {
  id: UUID;
  hierarchy_order: number;
}

export interface RoleReorderRequest {
  roles: RoleReorderItem[];
}

export type RoleList = RoleDetail[];